File: solutions/vbnet/matching-brackets/1/MatchingBrackets.vb
Imports System.Collections.Generic

Public Module MatchingBrackets
    Public Function IsPaired(ByVal input As String) As Boolean
        Dim pairs As New Dictionary(of String, String) From {
          {"(", ")"},
          {"[", "]"},
          {"{", "}"}
        }

        Dim stack as Stack(Of String) = New Stack(Of String)()
        For Each c as Char in input
          Dim s = c.ToString()
          If "{[(".Contains(s) Then
            stack.Push(s)
          ElseIf "}])".Contains(s) Then
            If stack.Count = 0 Then Return False
            Dim p = pairs(stack.Pop())
            If s <> p Then Return False
          End If
        Next
        Return stack.Count = 0
    End Function
End Module

File: solutions/csharp/matching-brackets/1/MatchingBrackets.cs
using System;
using System.Collections.Generic;

public static class MatchingBrackets
{
    public static bool IsPaired(string input)
    {
        var pairs = new Dictionary<string, string>
        {
            {"(", ")"},
            {"[", "]"},
            {"{", "}"},
        };
        var stack = new Stack<string>();

        foreach(var c in input)
        {
            var s = c.ToString();
            if ("{[(".Contains(s))
            {
                stack.Push(s);
            } 
            else if ("}])".Contains(s))
            {
                if (stack.Count == 0) return false;
                var p = pairs[stack.Pop()];
                if (s != p) return false;
            }
        }

        return stack.Count == 0;
    }
}
